python/rikai/spark/sql/codegen/mlflow_registry.py

```diff
@@ -0,0 +1,192 @@
+        validate: bool = True,
+    ):
+        self._run = run
+        spec = self._run_to_spec_dict()
+        spec.setdefault("options", {})
+        if options:
+            spec["options"].update(options)
+        super().__init__(spec, validate=validate)
+
+    @property
+    def run_id(self):
+        return self._run.info.run_id
+
+    @property
+    def start_time(self):
+        return self._run.info.start_time
+
+    @property
+    def end_time(self):
+        return self._run.info.end_time
+
+    @property
+    def artifact_root_uri(self):
+        return self._run.info.artifact_uri
+
+    def _run_to_spec_dict(self):
+        """Convert the Run into a ModelSpec
+
+        Returns
+        -------
+        spec: Dict[str, Any]
+        """
+        tags = self._run.data.tags
+        spec = {
+            "version": tags["rikai.spec.version"],
+            "schema": tags["rikai.output.schema"],
+            "model": {"flavor": tags["rikai.model.flavor"]},
+        }
+
+        # transforms
+        transforms = {}
+        if "rikai.transforms.pre" in tags:
+            transforms["pre"] = tags["rikai.transforms.pre"]
+        if "rikai.transforms.post" in tags:
+            transforms["post"] = tags["rikai.transforms.post"]
+        if len(transforms) > 0:
+            spec["transforms"] = transforms
+
+        # options
+        options = dict(self._run.data.params or {})  # for training
+        for key, value in tags.items():
+            key = key.lower().strip()
+            if key.startswith("rikai.option."):
+                options[key[len("rikai.option.") :]] = value
+        if len(options) > 0:
+            spec["options"] = options
+
+        # uri (to the model artifact)
+        if "rikai.model.artifact.uri" in tags:
+            spec["model"]["uri"] = tags["rikai.model.artifact.uri"]
+        else:
+            # /model/MLModel is currently the MLFlow model schema
+            spec["model"]["uri"] = os.path.join(
+                self._run.info.artifact_uri, "model/MLModel"
+            )
+        return spec
+
+
+def codegen_from_runid(
+    spark: SparkSession,
+    run_id: str,
+    name: Optional[str] = None,
+    options: Optional[Dict[str, str]] = None,
+) -> str:
+    """Generate code from an MLFlow runid
+
+    Parameters
+    ----------
+    spark : SparkSession
+        A live spark session
+    run_id : str
+        the mlflow runid corresponding to the model spec
+    name : str
+        The name of the model in the catalog
+    options : dict
+        Optional parameters passed to the model.
+
+    Returns
+    -------
+    str
+        Spark UDF function name for the generated data.
+    """
+    try:
+        import mlflow
+    except ImportError:
+        raise ImportError(
+            "Couldn't import mlflow. Please make sure to "
+            "`pip install mlflow` explicitly or install "
+            "the correct extras like `pip install rikai[all]`"
```

Comment:
`rikai[all]` is so big now, maybe we should just suggest to do `pip install rikai[mlflow]`?

Reply:
sg changed.